### PR TITLE
Adding Jenkins

### PR DESCRIPTION
**Pull request by uvvpavel**

### CODE REVIEW

File: lib_xcore_math/src/bfp/bfp_s16.c [hunk 1]
```diff
@@ -339,12 +339,10 @@ float bfp_s16_mean(
     headroom_t hr = HR_S32(sum) + 32;
     int64_t sum64 = ((int64_t)sum) << hr;
     int64_t mean64 = sum64 / ((int)b->length);
-    right_shift_t shr = MAX(0, 48 - HR_S64(mean64));
+    right_shift_t shr = MAX(0, 32 - HR_S64(mean64));
```
Comment (lucianomartin) [inline lib_xcore_math/src/bfp/bfp_s16.c:342]: I suppose this was the bug you found, well done!

File: lib_xcore_math/src/bfp/bfp_s16.c [hunk 2]
```diff
@@ -339,12 +339,10 @@ float bfp_s16_mean(
...
 
-    //TODO: astew: there's no reason to force the precision down to 16 bits after
-    //             getting rid of float_s16_t because s32_to_f32 handles 32 bits
     if(shr > 0) 
-        mean64 += 1 << (shr-1);
```
Comment (lucianomartin) [inline lib_xcore_math/src/bfp/bfp_s16.c:347]: can you add a comment to explain why you have to cast a uint32_t value into uint64_t?

File: lib_xcore_math/src/bfp/bfp_s32.c [hunk 3]
```diff
@@ -335,8 +335,8 @@ float_s32_t bfp_s32_mean(
     right_shift_t shr = MAX(0, 32 - HR_S64(mean));
 
     if(shr > 0)
-        mean += 1 << (shr-1);
-    
+        mean += (uint32_t)(1 << (shr-1));
```
Comment (lucianomartin) [inline lib_xcore_math/src/bfp/bfp_s32.c:338]: can you add a comment to explain why you have to cast a uint32_t value into uint64_t?

File: test/dct_tests/src/test_dctXX_inverse.c [hunk 4]
```diff
@@ -27,20 +27,13 @@ TEST_SETUP(dctXX_inverse) { fflush(stdout); }
...
 
-
 #define MAX_PROC_FRAME_LENGTH_LOG2 9
 #define MAX_PROC_FRAME_LENGTH (1<<MAX_PROC_FRAME_LENGTH_LOG2)
 
 #define MIN_DCT_N_LOG2  (3)
 #define LOOPS_LOG2  (0)
 
 
-
```
Comment (lucianomartin) [inline test/dct_tests/src/test_dctXX_inverse.c:38]: why is SMOKE_TEST not used here?

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+                  }
+                }
+              }
+            } // Build
+
+            stage('Unit tests xs3a') {
+              steps {
+                dir('lib_xcore_math/build_xs3a/test') {
+                  withTools(params.TOOLS_VERSION) {
+                    sh 'xrun --xscope --id 0 --args bfp_tests/bfp_tests.xe        -v'
```
Comment (lucianomartin) [inline Jenkinsfile:66]: what does "-v" option do? Doesn't it just print the version? If you want to increase the verbosity, you need to use "--verbose", but I never used it myself,

File: test/dct_tests/src/test_dctXX_inverse.c (hunk 4 above)
Comment (lucianomartin) [inline test/dct_tests/src/test_dctXX_inverse.c:38]: why is SMOKE_TEST not used here? Is it because `LOOPS_LOG2`  is already 0?

File: Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+  agent none
+
+  parameters {
+    string(
+      name: 'TOOLS_VERSION',
+      defaultValue: '15.2.1',
+      description: 'The XTC tools version'
+    )
+    booleanParam(
+      name: 'XMATH_SMOKE_TEST',
```
Comment (lucianomartin) [inline Jenkinsfile:19]: how long the tests last if XMATH_SMOKE_TEST is set to False? I wonder we should use it by default if they last less than 30 min. If we decide to run the tests in smoke test mode, when do we run the full tests?

File: Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+  options {
+    skipDefaultCheckout()
+    timestamps()
+    buildDiscarder(xmosDiscardBuildSettings())
+  } // options
+
+  stages {
+    stage('Bullds and tests') {
+      parallel {
+        stage('Linux builds and tests') {
```
Comment (lucianomartin) [inline Jenkinsfile:33]: why don't we parallelize the xs3 tests with the x86 ones? We will reduce the duration of the Jenkins run by a considerable amount.

**Review by lucianomartin — COMMENTED** [review]
Nice work! I like it is tidy and well commented. I have a few questions. The only important one is about parallelizing the tests.

**Review by lucianomartin — COMMENTED** [review]
One more comment.

File: lib_xcore_math/src/bfp/bfp_s16.c (hunk 1 above)
Comment (uvvpavel) [reply lib_xcore_math/src/bfp/bfp_s16.c:342]: This was not a bug but, it's resolving a TODO from below.

File: test/dct_tests/src/test_dctXX_inverse.c (hunk 4 above)
Comment (uvvpavel) [reply test/dct_tests/src/test_dctXX_inverse.c:38]: Yes, we can't go any lower

File: Jenkinsfile (hunk 7 above)
Comment (uvvpavel) [reply Jenkinsfile:33]: The way it's designed right now is that the smoke run takes 4 min, and the non-smoke run takes 30 min. This whole stage is executed on a single agent. It's not gonna bring much benefit for a smoke run. and will reduce a non-smoke run by 4 min as xs3 takes 27 min and x86 takes 4 min.

File: Jenkinsfile (hunk 5 above)
Comment (uvvpavel) [reply Jenkinsfile:66]: `-v` here is for the test itself, not `xrun`, hence it has `--args`. It means verbose

File: lib_xcore_math/src/bfp/bfp_s16.c (hunk 2 above)
Comment (uvvpavel) [reply lib_xcore_math/src/bfp/bfp_s16.c:347]: I did a slight update on this line. I don't really know how to comment it as it's a standard procedure for rounding

File: lib_xcore_math/src/bfp/bfp_s32.c (hunk 3 above)
Comment (uvvpavel) [reply lib_xcore_math/src/bfp/bfp_s32.c]: I did a slight update on this line. I don't really know how to comment it, as it's a standard procedure for rounding

File: Jenkinsfile (hunk 7 above)
Comment (lucianomartin) [reply Jenkinsfile:33]: Thanks for the explanation and the details

**Review by lucianomartin — CHANGES_REQUESTED** [review]
The Jenkins artifacts contain the build files for the documentation. They shouldn't be archived.
See http://srv-bri-jcim0:8080/blue/organizations/jenkins/XMOS%2Flib_xcore_math/detail/PR-150/18/artifacts

**Review by BrennanGit — CHANGES_REQUESTED** [review]

**Comment by BrennanGit** [comment]
> The Jenkins artifacts contain the build files for the documentation. They shouldn't be archived. See http://srv-bri-jcim0:8080/blue/organizations/jenkins/XMOS%2Flib_xcore_math/detail/PR-150/18/artifacts

This is basically because the docs build "fails". That means it bombs out before the cleanup of build files. I hope to fix this in the v4 modification.

**Comment by uvvpavel** [comment]
> > The Jenkins artifacts contain the build files for the documentation. They shouldn't be archived. See http://srv-bri-jcim0:8080/blue/organizations/jenkins/XMOS%2Flib_xcore_math/detail/PR-150/18/artifacts
> 
> This is basically because the docs build "fails". That means it bombs out before the cleanup of build files. I hope to fix this in the v4 modification.

@lucianomartin I'll leave it there for now as it's being archived with the rest of the docs.

**Review by lucianomartin — APPROVED** [review]
All good. The doc build artifacts will be removed in a future PR.

**Review by BrennanGit — APPROVED** [review]
Jenkins stuff looks good to me 👍

**Pull request merged**